src/chainstate/stacks/db/blocks.rs

```diff
@@ -2702,6 +2747,103 @@ impl StacksChainState {
 
         Ok(ret)
     }
+
+    pub fn will_admit_mempool_tx<R: Read>(&mut self, current_burn: &BurnchainHeaderHash, current_block: &BlockHeaderHash, fd: &mut R) -> Result<StacksTransaction, MemPoolRejection> {
```

Comment:
What is our line of thinking on mempool admissions?
With the current approach, the checks are touching the storage on multiple points. Should we be afraid of peers trying to DoS'ing a node by flooding it with expensive malformed transactions?
Where do we draw the line "I'm rejecting this transaction upfront" vs "I'm going to handle this invalid transaction and collect its fees"?

Reply:
I think there's a couple of ways to draw the line, but I believe the overarching rule is:

"Don't do anything in the checks if the costs would _vary_ dependent on input."

While different transaction _types_ will have different costs to check, no contract-call is more expensive than any other contract-call to check (the argument type check is linear with the size of the transaction, which itself is fixed size). Checking the argument types on contract-call's _may_ be problematic, and we can remove that check from the mempool admission checks if that turns out to be the case, but I believe that the dominating cost here is going to be the cost of the database lookups.

The database lookups are unavoidable however -- we cannot accept transactions into the mempool if they don't have a valid nonce, nor can we accept transactions from accounts without funds.